### PR TITLE
Plugin loader

### PR DESCRIPTION
**Pull request by dadgar**
This PR provides a plugin loader that scans a plugin directory for plugins and provides a unified interface for launching and reattaching to plugins whether they are internal or external.

The loader will configure the plugin before handing the instance back to the caller. The unit tests are slightly complicated, but the way it works is that there is a plugin main which allows switching to running the binary as a mock plugin. The plugin currently only can be a device plugin but can be extended later.

The test harness creates a plugin directory and then copies itself to the temp folder. This provides binaries that can be fingerprinted and run as plugins.

### CODE REVIEW

File: plugins/shared/loader/init.go [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+func (l *PluginLoader) scan() ([]os.FileInfo, error) {
+	// Capture the list of binaries in the plugins folder
+	files, err := ioutil.ReadDir(l.pluginDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read plugin directory %q: %v", l.pluginDir, err)
+	}
+
+	var plugins []os.FileInfo
+	for _, f := range files {
+		if f.IsDir() {
```
[the remark this answers is not in the document]
Comment (schmichael) [reply plugins/shared/loader/init.go]: Need to test if `f` is a symlink and if so resolve it and test if its a directory.

(Or use Readdirnames and Stat which follows symlinks for you.)

File: plugins/shared/loader/init.go [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+func (l *PluginLoader) fingerprintPlugins(plugins []os.FileInfo, configs map[string]*config.PluginConfig) (map[PluginID]*pluginInfo, error) {
+	var mErr multierror.Error
+	fingerprinted := make(map[PluginID]*pluginInfo, len(plugins))
+	for _, p := range plugins {
+		name := cleanPluginExecutable(p.Name())
+		c := configs[name]
+		info, err := l.fingerprintPlugin(p, c)
+		if err != nil {
+			l.logger.Error("failed to fingerprint plugin", "plugin", name)
+			multierror.Append(&mErr, err)
```
Comment (schmichael) [inline plugins/shared/loader/init.go]: So will Nomad fail to start if a plugin fails to fingerprint?

I think this is the correct behavior as it seems really bad to allow Nomad to run without a driver plugin that may be necessary to communicate with running tasks. An operator can always fix these errors by removing the offending file (or removing the execute bit if you make that change).

Just wanted to verify as this seems like an important decision.

File: plugins/shared/loader/init.go [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+			Name:       info.baseInfo.Name,
+			PluginType: info.baseInfo.Type,
+		}
+
+		// Detect if we already have seen a version of this plugin
+		if prev, ok := fingerprinted[id]; ok {
+
+			// Determine if we should keep the previous version or override
+			if prev.version.GreaterThan(info.version) {
+				l.logger.Info("multiple versions of plugin detected", "plugin", info.baseInfo.Name)
```
Comment (schmichael) [inline plugins/shared/loader/init.go]: Log both versions, filenames, and which version is being used.

File: plugins/shared/loader/init.go [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to get plugin info for plugin %q: %v", info.exePath, err)
+	}
+	info.baseInfo = i
+
+	// Parse and set the plugin version
+	if v, err := version.NewVersion(i.PluginVersion); err != nil {
+		return nil, fmt.Errorf("failed to parse plugin %q (%v) version %q: %v",
+			i.Name, info.exePath, i.PluginVersion, err)
+	} else {
```
Comment (schmichael) [inline plugins/shared/loader/init.go]: No need for explicit else

File: plugins/shared/loader/init.go [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+
+	for k, extPlugin := range external {
+		internal, ok := finalized[k]
+		if ok {
+			// We have overlapping plugins, determine if we should keep the
+			// internal version or override
+			if extPlugin.version.LessThan(internal.version) {
+				l.logger.Info("preferring external version of plugin",
+					"plugin", extPlugin.baseInfo.Name, "internal_version", internal.version.String(),
+					"external_version", extPlugin.version.String())
```
Comment (schmichael) [inline plugins/shared/loader/init.go]: No need to explicitly call `String()`. Anything that implements String() will have it called automatically.

File: plugins/shared/loader/loader.go [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+}
+
+// NewPluginLoader returns an instance of a plugin loader or an error if the
+// plugins could not be loaded
+func NewPluginLoader(config *PluginLoaderConfig) (*PluginLoader, error) {
+	if err := validateConfig(config); err != nil {
+		return nil, fmt.Errorf("invalid plugin loader configuration passed: %v", err)
+	}
+
+	logger := config.Logger.Named("plugin-loader").With("plugin-dir", config.PluginDir)
```
Comment (schmichael) [inline plugins/shared/loader/loader.go]: Hm, I've been using underscores as a separator in log names/keys so it looks like an identifier and matches hcl params (when applicable): https://gist.github.com/schmichael/da63914230b5127fb26486b01446f5a5

I don't mind switching, but we should probably standardize.

File: plugins/shared/loader/loader_test.go [hunk 7]
```diff
@@ -0,0 +1,851 @@
...
+
+// harness is used to build a temp directory and copy our own test executable
+// into it, allowing the plugin loader to scan for plugins.
+type harness struct {
+	t      *testing.T
+	tmpDir string
+}
+
+// newHarness returns a harness and copies our test binary to the temp directory
+// with teh passed plugin names.
```
Comment (schmichael) [inline plugins/shared/loader/loader_test.go]: the*

File: plugins/shared/loader/loader_test.go [hunk 8]
```diff
@@ -0,0 +1,851 @@
...
+	// Create two plugins
+	plugins := []string{"mock-device", "mock-device-2"}
+	pluginVersions := []string{"v0.0.1", "v0.0.2"}
+	h := newHarness(t, plugins)
+	defer h.cleanup()
+
+	logger := log.Default()
+	logger.SetLevel(log.Trace)
+	lconfig := &PluginLoaderConfig{
+		Logger:    logger, // XXX Use testlog package
```
Comment (schmichael) [inline plugins/shared/loader/loader_test.go]: :+1: my branch has this

File: plugins/shared/loader/init.go (hunk 2 above)
Comment (dadgar) [reply plugins/shared/loader/init.go]: Yes the New method to retrieve the factory which will happen in agent.go will return an error.

**Pull request merged**